### PR TITLE
feat: support delta writer

### PR DESCRIPTION
**Pull request by ZENOTME**

### CODE REVIEW

File: icelake/src/io/upsert_writer.rs [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+                        format!("Invalid ops: {op}"),
+                    ))
+                }
+            }
+        }
+        Ok(())
+    }
+
+    pub async fn current_write_metrics() -> Result<()> {
+        todo!()
```
Comment (liurenjie1024) [inline icelake/src/io/upsert_writer.rs:123]: Will this be left to following pr?

File: icelake/src/io/upsert_writer.rs [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+};
+
+use super::{
+    file_writer::{DeltaWriterResult, EqualityDeltaWriter},
+    location_generator::FileLocationGenerator,
+    DefaultFileAppender, FileAppenderBuilder, FileAppenderLayer,
+};
+use arrow_ord::partition::partition;
+
+pub enum UpsertWriter<L: FileAppenderLayer<DefaultFileAppender>> {
```
Comment (liurenjie1024) [inline icelake/src/io/upsert_writer.rs:26]: Is it possible to abstract this using generics? e.g. sth like
```
UpsertWriter<P: Partitioner, L: FileAppender>>
```

**Review by liurenjie1024 — COMMENTED** [review]
Others LGTM

File: icelake/src/io/upsert_writer.rs (hunk 2 above)
Comment (ZENOTME) [reply icelake/src/io/upsert_writer.rs:26]: Do you mean the user can specify partitioned or unpartitioned by `P: Partitioner`? Or different PartitionSplitter?

Comment (liurenjie1024) [reply icelake/src/io/upsert_writer.rs:26]: > Do you mean the user can specify partitioned or unpartitioned by P: Partitioner? 

I mean this. Unpartitioned is just a special partitioning which always returns same partition key. I'm just guessing with this approach we can reduce some duplicated codes. But it's an optimization, and up to you on how to implement it.

Comment (ZENOTME) [reply icelake/src/io/upsert_writer.rs:26]: Oh, I get it. But for now, we directly use EqualityDeltaWriter as an Unpartitioned writer. So it didn't have duplicated codes between partitioned and unpartitioned writer.

**Review by liurenjie1024 — APPROVED** [review]

**Pull request merged**